### PR TITLE
[MM-41733] Add Setup Wizard

### PR DESCRIPTION
**Pull request by hanzei**
#### Summary
This is a draft of the MVP Setup Wizard

Webhook verification is not possible, as webhooks can't be triggered via the API (https://gitlab.com/gitlab-org/gitlab/-/issues/25329)

#### Discussions
There are two ongoing discussion on the communit server:
- https://community-daily.mattermost.com/core/pl/nipu7obozfn87xoxx5fou8ah7o
- https://community-daily.mattermost.com/core/pl/br6zgtn9m7ydzrsg77abxsyucr

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-41734
https://mattermost.atlassian.net/browse/MM-41735
https://mattermost.atlassian.net/browse/MM-41736

### CODE REVIEW

**Comment by aaronrothschild** [comment]
@hanzei Is there a spinwick/etc I can try this out on?

**Comment by hanzei** [comment]
@aaronrothschild Yes: https://community-daily.mattermost.com/private-core/pl/zyakuqog6jnhfrbpagfghkgmwc

**Comment by Rina-dsg** [comment]
@hanzei Looks great! Just a few notes:
1. Confirmation message doesn't show @ name and contains btns.
Screenshot 2022-02-11 at 18 43 25
2. Not clear to which channel a message was sent
Screenshot 2022-02-11 at 19 12 05
3. The link seems broken.
Screenshot 2022-02-11 at 19 15 29

**Comment by hanzei** [comment]
@Rina-dsg Thanks for catching these bug :+1: , they should be fixed now. Please take another look.

**Comment by Rina-dsg** [comment]
1. Can change the style of the warning?
Screenshot 2022-02-15 at 18 12 03

**Comment by hanzei** [comment]
@Rina-dsg Unfortunately, there is not.

File: server/cluster.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	case oauthCompleteEventID:
+		var event OAuthCompleteEvent
+		if err := json.Unmarshal(ev.Data, &event); err != nil {
+			p.API.LogWarn("cannot unmarshal cluster event with OAuth complete event", "error", err)
+			return
+		}
+
+		p.oauthBroker.publishOAuthComplete(event.UserID, event.Err, true)
+	default:
+		p.API.LogWarn("unknown cluset event", "id", ev.Id)
```
Comment (catalintomai) [inline server/cluster.go:53]: cluset->cluster

File: server/command.go [hunk 2]
```diff
@@ -126,13 +132,43 @@ func (p *Plugin) ExecuteCommand(_ *plugin.Context, args *model.CommandArgs) (*mo
...
+
+	if err := config.IsValid(); err != nil {
+		isSysAdmin, err := p.isAuthorizedSysAdmin(args.UserId)
+		var text string
+		switch {
+		case err != nil:
+			text = "Error checking user's permissions"
+			p.API.LogWarn(text, "error", err.Error())
+		case isSysAdmin:
+			text = "Before using this plugin, you'll need to configure it by running `/github setup`"
```
Comment (catalintomai) [inline server/command.go:156]: github -> gitlab

File: server/command.go [hunk 3]
```diff
@@ -126,13 +132,43 @@ func (p *Plugin) ExecuteCommand(_ *plugin.Context, args *model.CommandArgs) (*mo
...
+		default:
+			text = "Please contact your system administrator to configure the GitHub plugin."
```
Comment (catalintomai) [inline server/command.go:158]: Github -> Gitlab

File: server/configuration.go [hunk 4]
```diff
@@ -124,15 +190,45 @@ func (p *Plugin) OnConfigurationChange() error {
...
 	}
 
+	enableDiagnostics := false
+	if config := p.API.GetConfig(); config != nil {
+		if configValue := config.LogSettings.EnableDiagnostics; configValue != nil {
+			enableDiagnostics = *configValue
+		}
+	}
+
+	p.tracker = telemetry.NewTracker(p.telemetryClient, p.API.GetDiagnosticId(), p.API.GetServerVersion(), manifest.Id, manifest.Version, "github", enableDiagnostics)
```
Comment (catalintomai) [inline server/configuration.go:216]: github -> gitlab

File: server/subscriptions.go [hunk 5]
```diff
@@ -141,7 +143,8 @@ func (p *Plugin) GetSubscribedChannelsForProject(
 
 	subsToReturn := make([]*subscription.Subscription, 0, len(subsForRepo))
 	for _, sub := range subsForRepo {
-		if !isPublicVisibility && !p.permissionToProject(sub.CreatorID, namespace, project) {
+		// TODO: Add context
```
Comment (catalintomai) [inline server/subscriptions.go:146]: need to open ticket?

**Review by catalintomai — APPROVED** [review]

**Review by Rina-dsg — APPROVED** [review]

File: server/flow.go [hunk 6]
```diff
@@ -0,0 +1,825 @@
...
+		"4. Select `api` and `read_user` in Scopes.\n"+
+		"5. Select **Save application**\n",
+		fm.pluginURL,
+	)
+
+	return flow.NewStep(stepOAuthInfo).
+		WithPretext(oauthPretext).
+		WithText(oauthMessage).
+		//TODO: Add image
+		// WithImage(fm.pluginURL, "public/new-oauth-application.png").
```
Comment (hanzei) [inline server/flow.go:451]: @Rina-dsg Could you please provide a screenshot for the "Create Application" step?

Comment (hanzei) [reply server/flow.go]: It looks like this right now. Please let me know what you think.
Screenshot from 2022-02-24 11-51-53
Screenshot from 2022-02-24 11-51-59

**Comment by hanzei** [comment]
@aaronrothschild @levb Gentle reminder to review this PR

**Review by levb — APPROVED** [review]

**Comment by DHaussermann** [comment]
1. I'm seeing an issue where GitLab authentication is not working. I'm still investigating but these are the details as I understand them now. 
- When using your own oAuth App the plugin seems to work fine. 
- When using Chimera the plugin has issues
  - Sometimes authentication seems to work but sometimes I see this error
image
 - Even when the authentication flow **does** seem to work - There is no sign in the UI anything has gone wrong but, the plugin is not usable. It is behaving as though setup was not done. 
image
  
2. Webhook Creation is not working now. (Possibly only when using Chimera flow)
When trying to create the webhook, the plugin behaves as though it is misconfigured. 
image
The UI does indeed show that the necessary values where stored when the plugin was enabled as expected.
image


@hanzei if you have a bit of time can you share your thoughts? It seems like possibly all issues above could be caused by the plugin not being able to read the values that it is writing to the config. Is it possible that such a problem is specific to Chimera flow? Also strange that none of these issues occurred for GitHub but, I'll have another look.

Nest steps:
- Go back and test oAuth app branch of setup flow to see if anything is going wrong. 
- Test this locally to ensure the keys the plugin is looking for match verbatim what's being written to the config file
- Explore GitHub in case any of these issues may reproduce there.

**Comment by DHaussermann** [comment]
Follow up on the above:
- This **does** seem to only occur with using the Chimera branch.
- The config keys look normal and update as expected when this is tested locally (no key duplication or inconsistency with key names) but - I can't complete a full test as I don't have the proxy available
- GitHub does **not** show any of the issues when retesting

**Comment by hanzei** [comment]
@DHaussermann thanks for the in depth review!

Before I try to reproduce the issue myself: Does the authentication work with Chimera on `master`?

**Comment by DHaussermann** [comment]
@hanzei Deploying the master branch and authenticating with Chimera once it's enabled does not show the connection issue. Also, the plugin is functional and does not tell me I needs to be configured when I make use of it.

**Comment by hanzei** [comment]
@DHaussermann I can't reproduce any of the issues that you pointed out above. I've tested with a local setup and configured a Chimera URL. Connecting the account works fine and the plugin is usable after the setup completes.

Is there anything related in the server logs?

**Comment by DHaussermann** [comment]
@dipak-demansol can you help by providing another data point here?

Steps:
- Setup cloud test server (this will have proxy pre-configured)
- Install and enable the plugin (Dylan can provide a build)
- Set Chimera proxy to true and Save 
- Run `/gitlab setup`
- Follow setup flow that does not involve creating new oAuth app

Please advise on if you can reproduce issues above.

**Comment by dipak-demansol** [comment]
@DHaussermann @hanzei here is the video to regenerate the Issue. https://drive.google.com/file/d/1lHZAj_uFb4hfUnNVSvwycwRd7Dots5Gc/view?usp=sharing

**Comment by hanzei** [comment]
@dipak-demansol 

What commit of the GitLab plugin did you use for testing? Is there anything meaningful in the logs when the plugin starts?

Could it be that the encryption key is empty?

**Comment by dipak-demansol** [comment]
> @dipak-demansol
> 
> What commit of the GitLab plugin did you use for testing? Is there anything meaningful in the logs when the plugin starts?
> 
> Could it be that the encryption key is empty?

i don't know about the commit because dylan given me the plugin file, here is the log file https://drive.google.com/file/d/14RooHYVYJHjdl0GajA5E9Gg6VKR64P5h/view?usp=sharing

**Comment by hanzei** [comment]
@dipak-demansol Could you please check if the issue reproduces with the latest commit?

**Comment by dipak-demansol** [comment]
> @dipak-demansol Could you please check if the issue reproduces with the latest commit?

@hanzei i did testing with latest master build and it worked, issue not generated but when i was doing some experiment then in the jira i was getting this error while **Upload the Mattermost app to Jira** 
image
it can be a issue or we can ignore it?

**Comment by DHaussermann** [comment]
@hanzei I compiled the build for @dipak-demansol I seem to be running the latest commit
image
I can try this again as a quick sanity check and let you know.

**Comment by hanzei** [comment]
@DHaussermann Is there anything I should look into right now?

**Review by dipak-demansol — APPROVED** [review]
LGTM, below is the list of tested points.
      :- With Chimera i'm able to Complete the Configuration by using the /setup command.
      :- after Chimera Configuration i regenerated the `Webhook Secret` & `At Rest Encryption Key` and by using /gitlab me i got proper validation message as "Before using this plugin, you'll need to configure it by running /gitlab Setup".
      :- /setup command With personal access token i'm able to do the Configuration. 
      :- earlier we found a issue where the `Webhook Secret` & `At Rest Encryption Key` data was missing from the input box after enable the plugin but now that issue is not regenerated, tested with 2 different server.
       :- While doing Chimera Configuration i got many auto-suggested command but when Chimera is OFF then only got `Setup` Command so it's good.


non blocking point already requested by dylan.
----------------------------- 
this form takes id password automatically.
image

**Review by DHaussermann — APPROVED** [review]
Thanks for testing @dipak-demansol. 

The issue is not reproducible.   
Created a separate issue to address the remaining concern here https://github.com/mattermost/mattermost-plugin-gitlab/issues/286

**Comment by hanzei** [comment]
Blocking until the discussion around https://mattermost.atlassian.net/browse/MM-42918 is resolved

**Comment by hanzei** [comment]
https://mattermost.atlassian.net/browse/MM-42918 will need to be resolved also for this plugin. Let's merge this PR, work on a solution for https://mattermost.atlassian.net/browse/MM-42918 and the apply it to this plugin.

**Pull request merged**